In [37]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from pathlib import Path

In [38]:
# Вывод данных из csv во всю ширину окна
pd.set_option('display.max_rows', 500)
pd.set_option('display.max_columns', 500)
pd.set_option('display.width', 1000)

## Убираем индекс

In [39]:
df0 = pd.DataFrame({"ID": ["A", "B", "A", "C", "A", "A", "C", "B"], "value": [1, 2, 4, 3, 6, 7, 3, 4]})
df_group1 = df0.groupby("ID").sum().reset_index()
df_group1
df_group2 = df0.groupby("ID", as_index=False).sum()
df_group2

ID  value
0  A     18
1  B      6
2  C      6

1) Имеются следующие данные о транзакциях в период с 01.12.2010 по 12.09.2011:
InvoiceNo – номер транзакции
StockCode – код товара
Description – описание товара
Quantity – количество единиц товара, добавленных в заказ
InvoiceDate – дата транзакции
UnitPrice – цена за единицу товара
CustomerID – id клиента
Country – страна, где проживает клиент
Импортируйте данные data.csv.zip (в папке) с кодировкой ISO-8859-1. Запишите полученный датафрейм в retail, а названия колонок сохраните в переменную retail_columns.

In [40]:
# Импортируйте данные data.csv.zip (в папке) с кодировкой ISO-8859-1. Запишите полученный датафрейм в retail,
# а названия колонок сохраните в переменную retail_columns.

path = 'e:\!Data-scientist\!!LessonPre\[Stepic] Анатолий Карпов - Твой путь в мир Data Analytics\Data\data5.csv.zip'
retail = pd.read_csv(path,
            encoding='ISO-8859-1',
            compression='zip'
        )
print('Импортируем данные data.csv.zip (в папке) с кодировкой ISO-8859-1.\nКоличество строк и колонок:', retail.shape)

retail_columns = retail.columns
print('\nНазвания колонок сохраним в переменную retail_columns:\n', retail_columns)
print('\nПервые 5 строк:\n', retail.head())

Импортируем данные data.csv.zip (в папке) с кодировкой ISO-8859-1.
Количество строк и колонок: (541909, 8)

Названия колонок сохраним в переменную retail_columns:
 Index(['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate', 'UnitPrice', 'CustomerID', 'Country'], dtype='object')

Первые 5 строк:
   InvoiceNo StockCode                          Description  Quantity     InvoiceDate  UnitPrice  CustomerID         Country
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6  12/1/2010 8:26       2.55     17850.0  United Kingdom
1    536365     71053                  WHITE METAL LANTERN         6  12/1/2010 8:26       3.39     17850.0  United Kingdom
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8  12/1/2010 8:26       2.75     17850.0  United Kingdom
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6  12/1/2010 8:26       3.39     17850.0  United Kingdom
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6

2) Проверьте, встречаются ли в данных повторяющиеся наблюдения, и в качестве ответа укажите их количество. Если они есть, то удалите их из retail.

In [41]:
nrow_original = retail.shape[0]
print('\nПервично было:', nrow_original)


Первично было: 541909


In [42]:
retail = retail.drop_duplicates()
print('\nПроверяем данные на дупликаты и удаляем повторяющиеся:', retail.shape[0])


Проверяем данные на дупликаты и удаляем повторяющиеся: 536641


3) ⭐️Задание со звёздочкой!⭐️
Данные содержат в себе записи как и об успешных транзакциях, так и об отмененных. Если пользователь отменил заказ, в начале номера транзакции (InvoiceNo) ставится C (canceled).
Сколько всего заказов отменили пользователи?

In [43]:
Invoice = retail.InvoiceNo.\
    str.startswith('C').sum()
print('\nВсего заказов отменили', Invoice, 'пользователей')


Всего заказов отменили 9251 пользователей


4) Теперь отфильтруйте данные и оставьте в retail только те заказы, где Quantity > 0. В качестве ответа укажите число оставшихся строк.

In [44]:
retail = retail.query('Quantity > 0')
print('\nЗаказы, где больше 0 единиц товара:', retail.shape[0])


Заказы, где больше 0 единиц товара: 526054


5) 💀 Сложное задание! 💀
Посчитайте число заказов для каждого пользователя (CustomerID) из Германии (Germany). Оставьте только тех, кто совершил более N транзакций (InvoiceNo), где N – 80% процентиль. Запишите полученные id пользователей в germany_top (не весь датафрейм, только id). Идентификатор заказа – InvoiceNo. Для каждого заказа в данных может встречаться более 1 строки.

In [45]:
germany_buyers = retail.query('Country == "Germany"') \
    .groupby('CustomerID', as_index=False)\
    .agg({'InvoiceNo': pd.Series.nunique})

percentile_80 = germany_buyers.InvoiceNo.quantile(q=0.8)
print('\nЧисло заказов  из Германии 80% процентиль:', percentile_80)

germany_top = germany_buyers\
    .rename(columns={'InvoiceNo': 'orders_numbers'})\
    .query('orders_numbers > @percentile_80')\
    .CustomerID
print('\nId пользователей из Германии, процентиль заказов 80% :', germany_top.tolist())


Число заказов  из Германии 80% процентиль: 7.0

Id пользователей из Германии, процентиль заказов 80% : [12471.0, 12474.0, 12476.0, 12481.0, 12500.0, 12524.0, 12569.0, 12600.0, 12619.0, 12621.0, 12626.0, 12647.0, 12662.0, 12705.0, 12708.0, 12709.0, 12712.0, 12720.0]


6) ⭐️Задание со звёздочкой!⭐️
Используя объект с id пользователей (germany_top), полученный на предыдущем шаге, отфильтруйте наблюдения и оставьте в данных записи только по интересующим нас юзерам. Результирующий датафрейм запишите в top_retail_germany.

In [46]:
top_retail_germany = retail.query('CustomerID in @germany_top')
top_retail_germany.reset_index(drop=True, inplace=True)  # Сброс индекса
print('\nНаблюдения за топ пользователями из Германии, процентиль заказов 80% :\n', top_retail_germany)


Наблюдения за топ пользователями из Германии, процентиль заказов 80% :
      InvoiceNo StockCode                          Description  Quantity      InvoiceDate  UnitPrice  CustomerID  Country
0       536527     22809              SET OF 6 T-LIGHTS SANTA         6  12/1/2010 13:04       2.95     12662.0  Germany
1       536527     84347  ROTATING SILVER ANGELS T-LIGHT HLDR         6  12/1/2010 13:04       2.55     12662.0  Germany
2       536527     84945   MULTI COLOUR SILVER T-LIGHT HOLDER        12  12/1/2010 13:04       0.85     12662.0  Germany
3       536527     22242        5 HOOK HANGER MAGIC TOADSTOOL        12  12/1/2010 13:04       1.65     12662.0  Germany
4       536527     22244           3 HOOK HANGER MAGIC GARDEN        12  12/1/2010 13:04       1.95     12662.0  Germany
...        ...       ...                                  ...       ...              ...        ...         ...      ...
4170    581570     22139     RETROSPOT TEA SET CERAMIC 11 PC          3  12/9/20

7) Сгруппируйте top_retail_germany по коду товара (StockCode). Какой из продуктов добавляли в корзину чаще всего, кроме POST?
Note: одним заказом считается единовременная покупка любого количества товара, т.е. без учета Quantity.

In [47]:
top_quantity_germany = top_retail_germany \
    .StockCode.value_counts()
print('\nПродукт добавляли в корзину чаще всего v1:', top_quantity_germany)

top_quantity_germany1 = top_retail_germany \
    .groupby('StockCode')\
    .agg({'Description': 'count'}) \
    .sort_values('Description', ascending=False)\
    .drop(index=['POST'])
print('\nПродукт добавляли в корзину чаще всего v2:\n', top_quantity_germany1)


Продукт добавляли в корзину чаще всего v1: POST     213
22326     52
22328     38
22423     34
20719     30
        ... 
21564      1
22584      1
22561      1
21497      1
23562      1
Name: StockCode, Length: 1158, dtype: int64

Продукт добавляли в корзину чаще всего v2:
            Description
StockCode             
22326               52
22328               38
22423               34
20719               30
22961               28
...                ...
22563                1
22569                1
22574                1
22576                1
22662                1

[1157 rows x 1 columns]


8) Вернемся к анализу полного датасета retail. Создайте колонку Revenue с суммой покупки, используя колонки Quantity и UnitPrice.

In [48]:
# retail['Revenue'] = retail.UnitPrice * retail.Quantity # 1й вариант

# 2й вариант
retail = retail.assign(
        Revenue = retail.UnitPrice * retail.Quantity
)
print(retail)

       InvoiceNo StockCode                          Description  Quantity      InvoiceDate  UnitPrice  CustomerID         Country  Revenue
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   12/1/2010 8:26       2.55     17850.0  United Kingdom    15.30
1         536365     71053                  WHITE METAL LANTERN         6   12/1/2010 8:26       3.39     17850.0  United Kingdom    20.34
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   12/1/2010 8:26       2.75     17850.0  United Kingdom    22.00
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   12/1/2010 8:26       3.39     17850.0  United Kingdom    20.34
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   12/1/2010 8:26       3.39     17850.0  United Kingdom    20.34
...          ...       ...                                  ...       ...              ...        ...         ...             ...      ...
541904    581587     22613 

9) Для каждой транзакции (InvoiceNo), посчитайте финальную сумму заказа. В качестве ответа укажите топ-5 (InvoiceNo) по сумме заказа (через запятую с пробелом и в порядке убывания TotalRevenue)

Например, для следующего примера...

InvoiceNo StockCode Quantity InvoiceDate          UnitPrice  CustomerID  Revenue
536365    85123A    6        2010-12-01 08:26:00  2.55       17850.0     15.30
536365    71053     6        2010-12-01 08:26:00  3.39       17850.0     20.34
555555    71053     1        2010-12-02 18:00:00  3.39       17850.0     3.39
...сумма заказа будет равна:

InvoiceNo  TotalRevenue
536365     35.64
555555     3.39

И ответом будет:
536365, 555555

In [49]:
retail_top = retail\
        .groupby('InvoiceNo', as_index=False)\
        .agg({'Revenue': 'sum'}) \
        .sort_values('Revenue',  ascending=False)
print('Топ-5 по сумме заказа. Через запятую с пробелом и в порядке убывания:', retail_top.InvoiceNo[:5].str.cat(sep=', '))

Топ-5 по сумме заказа. Через запятую с пробелом и в порядке убывания: 581483, 541431, 574941, 576365, 556444
